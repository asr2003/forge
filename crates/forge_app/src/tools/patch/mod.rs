mod apply;
mod apply_json;
mod marker;
mod parse;

pub use apply_json::ApplyPatchJson;
