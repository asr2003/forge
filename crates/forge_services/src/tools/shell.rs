// PathBuf now comes from the ShellInput in forge_domain
use std::sync::Arc;

use anyhow::bail;
use forge_display::TitleFormat;
use forge_domain::{
    CommandOutput, Environment, EnvironmentService, ExecutableTool, NamedTool, ShellInput,
    ToolCallContext, ToolDescription, ToolName,
};
use forge_tool_macros::ToolDescription;
use strip_ansi_escapes::strip;

use crate::metadata::Metadata;
use crate::{Clipper, ClipperResult, CommandExecutorService, FsWriteService, Infrastructure};

/// Number of characters to keep at the start of truncated output
const PREFIX_CHARS: usize = 10_000;

/// Number of characters to keep at the end of truncated output
const SUFFIX_CHARS: usize = 10_000;

// Using ShellInput from forge_domain

// Strips out the ansi codes from content.
fn strip_ansi(content: String) -> String {
    String::from_utf8_lossy(&strip(content.as_bytes())).into_owned()
}

/// Formats command output by wrapping non-empty stdout/stderr in XML tags.
/// stderr is commonly used for warnings and progress info, so success is
/// determined by exit status, not stderr presence. Returns Ok(output) on
/// success or Err(output) on failure, with a status message if both streams are
/// empty.
async fn format_output<F: Infrastructure>(
    infra: &Arc<F>,
    mut output: CommandOutput,
    keep_ansi: bool,
    prefix_chars: usize,
    suffix_chars: usize,
) -> anyhow::Result<String> {
    let mut formatted_output = String::new();

    if !keep_ansi {
        output.stderr = strip_ansi(output.stderr);
        output.stdout = strip_ansi(output.stdout);
    }

    // Create metadata
    let mut metadata = Metadata::default()
        .add("command", &output.command)
        .add_optional("exit_code", output.exit_code);

    let mut is_truncated = false;

    // Format stdout if not empty
    if !output.stdout.trim().is_empty() {
        let result = Clipper::from_start_end(prefix_chars, suffix_chars).clip(&output.stdout);

        if result.is_truncated() {
            metadata = metadata.add("total_stdout_chars", output.stdout.len());
            is_truncated = true;
        }
        formatted_output.push_str(&tag_output(result, "stdout", &output.stdout));
    }

    // Format stderr if not empty
    if !output.stderr.trim().is_empty() {
        if !formatted_output.is_empty() {
            formatted_output.push('\n');
        }
        let result = Clipper::from_start_end(prefix_chars, suffix_chars).clip(&output.stderr);

        if result.is_truncated() {
            metadata = metadata.add("total_stderr_chars", output.stderr.len());
            is_truncated = true;
        }
        formatted_output.push_str(&tag_output(result, "stderr", &output.stderr));
    }

    // Add temp file path if output is truncated
    if is_truncated {
        let path = infra
            .file_write_service()
            .write_temp(
                "forge_shell_",
                ".md",
                &format!(
                    "command:{}\n<stdout>{}</stdout>\n<stderr>{}</stderr>",
                    output.command, output.stdout, output.stderr
                ),
            )
            .await?;

        metadata = metadata
            .add("temp_file", path.display())
            .add("truncated", "true");
        formatted_output.push_str(&format!(
            "<truncate>content is truncated, remaining content can be read from path:{}</truncate>",
            path.display()
        ));
    }

    // Handle empty outputs
    let result = if formatted_output.is_empty() {
        if output.success() {
            "Command executed successfully with no output.".to_string()
        } else {
            "Command failed with no output.".to_string()
        }
    } else {
        formatted_output
    };

    if output.success() {
        Ok(format!("{metadata}{result}"))
    } else {
        bail!(format!("{metadata}{result}"))
    }
}

/// Helper function to format potentially truncated output for stdout or stderr
fn tag_output(result: ClipperResult, tag: &str, content: &str) -> String {
    let mut formatted_output = String::default();
    match (result.prefix, result.suffix) {
        (Some(prefix), Some(suffix)) => {
            let truncated_chars = content.len() - prefix.len() - suffix.len();
            let prefix_content = &content[prefix.clone()];
            let suffix_content = &content[suffix.clone()];

            formatted_output.push_str(&format!(
                "<{} chars=\"{}-{}\">\n{}\n</{}>\n",
                tag, prefix.start, prefix.end, prefix_content, tag
            ));
            formatted_output.push_str(&format!(
                "<truncated>...{tag} truncated ({truncated_chars} characters not shown)...</truncated>\n"
            ));
            formatted_output.push_str(&format!(
                "<{} chars=\"{}-{}\">\n{}\n</{}>\n",
                tag, suffix.start, suffix.end, suffix_content, tag
            ));
        }
        _ => formatted_output.push_str(&format!("<{tag}>\n{content}\n</{tag}>")),
    }

    formatted_output
}

/// Executes shell commands with safety measures using restricted bash (rbash).
/// Prevents potentially harmful operations like absolute path execution and
/// directory changes. Use for file system interaction, running utilities,
/// installing packages, or executing build commands. For operations requiring
/// unrestricted access, advise users to run forge CLI with '-u' flag. Returns
/// complete output including stdout, stderr, and exit code for diagnostic
/// purposes.
#[derive(ToolDescription)]
pub struct Shell<I> {
    env: Environment,
    infra: Arc<I>,
}

impl<I: Infrastructure> Shell<I> {
    /// Create a new Shell with environment configuration
    pub fn new(infra: Arc<I>) -> Self {
        let env = infra.environment_service().get_environment();
        Self { env, infra }
    }
}

impl<I> NamedTool for Shell<I> {
    fn tool_name() -> ToolName {
        ToolName::new("forge_tool_process_shell")
    }
}

#[async_trait::async_trait]
impl<I: Infrastructure> ExecutableTool for Shell<I> {
    type Input = ShellInput;

    async fn call(&self, context: ToolCallContext, input: Self::Input) -> anyhow::Result<String> {
        // Validate empty command
        if input.command.trim().is_empty() {
            bail!("Command string is empty or contains only whitespace".to_string());
        }
        let title_format = TitleFormat::debug(format!("Execute [{}]", self.env.shell.as_str()))
            .sub_title(&input.command);

        context.send_text(title_format).await?;

        let output = self
            .infra
            .command_executor_service()
            .execute_command(input.command, input.cwd)
            .await?;

        format_output(
            &self.infra,
            output,
            input.keep_ansi,
            PREFIX_CHARS,
            SUFFIX_CHARS,
        )
        .await
    }
}

#[cfg(test)]
mod tests {
    #[tokio::test]
    async fn test_format_output_with_different_max_chars() {
        let infra = Arc::new(MockInfrastructure::new());

        // Test with small truncation values that will truncate the string
        let small_output = CommandOutput {
            stdout: "ABCDEFGHIJKLMNOPQRSTUVWXYZ".to_string(),
            stderr: "".to_string(),
            command: "echo".into(),
            exit_code: Some(0),
        };
        let small_result = format_output(&infra, small_output, false, 5, 5)
            .await
            .unwrap();
        insta::assert_snapshot!(
            "format_output_small_truncation",
            TempDir::normalize(&small_result)
        );

        // Test with large values that won't cause truncation
        let large_output = CommandOutput {
            stdout: "ABCDEFGHIJKLMNOPQRSTUVWXYZ".to_string(),
            stderr: "".to_string(),
            command: "echo".into(),
            exit_code: Some(0),
        };
        let large_result = format_output(&infra, large_output, false, 100, 100)
            .await
            .unwrap();
        insta::assert_snapshot!(
            "format_output_no_truncation",
            TempDir::normalize(&large_result)
        );
    }
    use std::env;
    use std::sync::Arc;

    use pretty_assertions::assert_eq;

    use super::*;
    use crate::attachment::tests::MockInfrastructure;
    use crate::tools::utils::TempDir;

    /// Platform-specific error message patterns for command not found errors
    #[cfg(target_os = "windows")]
    const COMMAND_NOT_FOUND_PATTERNS: [&str; 2] = [
        "is not recognized",             // cmd.exe
        "'non_existent_command' is not", // PowerShell
    ];

    #[cfg(target_family = "unix")]
    const COMMAND_NOT_FOUND_PATTERNS: [&str; 3] = [
        "command not found",               // bash/sh
        "non_existent_command: not found", // bash/sh (Alternative Unix error)
        "No such file or directory",       // Alternative Unix error
    ];

    #[tokio::test]
    async fn test_shell_echo() {
        let infra = Arc::new(MockInfrastructure::new());
        let shell = Shell::new(infra);
        let result = shell
            .call(
                ToolCallContext::default(),
                ShellInput {
                    command: "echo 'Hello, World!'".to_string(),
                    cwd: env::current_dir().unwrap(),
                    keep_ansi: true,
                },
            )
            .await
            .unwrap();
        assert!(result.contains("Mock command executed successfully"));
    }

    #[tokio::test]
    async fn test_shell_stderr_with_success() {
        let infra = Arc::new(MockInfrastructure::new());
        let shell = Shell::new(infra);
        // Use a command that writes to both stdout and stderr
        let result = shell
            .call(
                ToolCallContext::default(),
                ShellInput {
                    command: if cfg!(target_os = "windows") {
                        "echo 'to stderr' 1>&2 && echo 'to stdout'".to_string()
                    } else {
                        "echo 'to stderr' >&2; echo 'to stdout'".to_string()
                    },
                    cwd: env::current_dir().unwrap(),
                    keep_ansi: true,
                },
            )
            .await
            .unwrap();
        insta::assert_snapshot!(result);
    }

    #[tokio::test]
    async fn test_shell_both_streams() {
        let infra = Arc::new(MockInfrastructure::new());
        let shell = Shell::new(infra);
        let result = shell
            .call(
                ToolCallContext::default(),
                ShellInput {
                    command: "echo 'to stdout' && echo 'to stderr' >&2".to_string(),
                    cwd: env::current_dir().unwrap(),
                    keep_ansi: true,
                },
            )
            .await
            .unwrap();
        insta::assert_snapshot!(result);
    }

    #[tokio::test]
    async fn test_shell_with_working_directory() {
        let infra = Arc::new(MockInfrastructure::new());
        let shell = Shell::new(infra);
        let temp_dir = TempDir::new().unwrap().path();

        let result = shell
            .call(
                ToolCallContext::default(),
                ShellInput {
                    command: if cfg!(target_os = "windows") {
                        "cd".to_string()
                    } else {
                        "pwd".to_string()
                    },
                    cwd: temp_dir.clone(),
                    keep_ansi: true,
                },
            )
            .await
            .unwrap();
        insta::assert_snapshot!(
            "format_output_working_directory",
            TempDir::normalize(&result)
        );
    }

    #[tokio::test]
    async fn test_shell_invalid_command() {
        let shell = Shell::new(Arc::new(MockInfrastructure::new()));
        let result = shell
            .call(
                ToolCallContext::default(),
                ShellInput {
                    command: "non_existent_command".to_string(),
                    cwd: env::current_dir().unwrap(),
                    keep_ansi: true,
                },
            )
            .await;

        assert!(result.is_err());
        let err = result.unwrap_err();

        // Check if any of the platform-specific patterns match
        let matches_pattern = COMMAND_NOT_FOUND_PATTERNS
            .iter()
            .any(|&pattern| err.to_string().contains(pattern));

        assert!(
            matches_pattern,
            "Error message '{err}' did not match any expected patterns for this platform: {COMMAND_NOT_FOUND_PATTERNS:?}"
        );
    }

    #[tokio::test]
    async fn test_shell_empty_command() {
        let infra = Arc::new(MockInfrastructure::new());
        let shell = Shell::new(infra);
        let result = shell
            .call(
                ToolCallContext::default(),
                ShellInput {
                    command: "".to_string(),
                    cwd: env::current_dir().unwrap(),
                    keep_ansi: true,
                },
            )
            .await;
        assert!(result.is_err());
        assert_eq!(
            result.unwrap_err().to_string(),
            "Command string is empty or contains only whitespace"
        );
    }

    #[tokio::test]
    async fn test_description() {
        assert!(
            Shell::new(Arc::new(MockInfrastructure::new()))
                .description()
                .len()
                > 100
        )
    }

    #[tokio::test]
    async fn test_shell_pwd() {
        let shell = Shell::new(Arc::new(MockInfrastructure::new()));
        let current_dir = env::current_dir().unwrap();
        let result = shell
            .call(
                ToolCallContext::default(),
                ShellInput {
                    command: if cfg!(target_os = "windows") {
                        "cd".to_string()
                    } else {
                        "pwd".to_string()
                    },
                    cwd: current_dir.clone(),
                    keep_ansi: true,
                },
            )
            .await
            .unwrap();

        assert_eq!(
            result,
            format!(
                "{}<stdout>\n{}\n\n</stdout>",
                Metadata::default()
                    .add(
                        "command",
                        if cfg!(target_os = "windows") {
                            "cd"
                        } else {
                            "pwd"
                        }
                    )
                    .add("exit_code", 0)
                    .to_string(),
                current_dir.display()
            )
        );
    }

    #[tokio::test]
    async fn test_shell_multiple_commands() {
        let shell = Shell::new(Arc::new(MockInfrastructure::new()));
        let result = shell
            .call(
                ToolCallContext::default(),
                ShellInput {
                    command: "echo 'first' && echo 'second'".to_string(),
                    cwd: env::current_dir().unwrap(),
                    keep_ansi: true,
                },
            )
            .await
            .unwrap();
        insta::assert_snapshot!(result);
    }

    #[tokio::test]
    async fn test_shell_empty_output() {
        let shell = Shell::new(Arc::new(MockInfrastructure::new()));
        let result = shell
            .call(
                ToolCallContext::default(),
                ShellInput {
                    command: "true".to_string(),
                    cwd: env::current_dir().unwrap(),
                    keep_ansi: true,
                },
            )
            .await
            .unwrap();

        assert!(result.contains("executed successfully"));
        assert!(!result.contains("failed"));
    }

    #[tokio::test]
    async fn test_shell_whitespace_only_output() {
        let shell = Shell::new(Arc::new(MockInfrastructure::new()));
        let result = shell
            .call(
                ToolCallContext::default(),
                ShellInput {
                    command: "echo ''".to_string(),
                    cwd: env::current_dir().unwrap(),
                    keep_ansi: true,
                },
            )
            .await
            .unwrap();

        assert!(result.contains("executed successfully"));
        assert!(!result.contains("failed"));
    }

    #[tokio::test]
    async fn test_shell_with_environment_variables() {
        let shell = Shell::new(Arc::new(MockInfrastructure::new()));
        let result = shell
            .call(
                ToolCallContext::default(),
                ShellInput {
                    command: "echo $PATH".to_string(),
                    cwd: env::current_dir().unwrap(),
                    keep_ansi: true,
                },
            )
            .await
            .unwrap();

        assert!(!result.is_empty());
        assert!(!result.contains("Error:"));
    }

    #[tokio::test]
    async fn test_shell_full_path_command() {
        let shell = Shell::new(Arc::new(MockInfrastructure::new()));
        // Using a full path command which would be restricted in rbash
        let cmd = if cfg!(target_os = "windows") {
            r"C:\Windows\System32\whoami.exe"
        } else {
            "/bin/ls"
        };

        let result = shell
            .call(
                ToolCallContext::default(),
                ShellInput {
                    command: cmd.to_string(),
                    cwd: env::current_dir().unwrap(),
                    keep_ansi: true,
                },
            )
            .await;

        // In rbash, this would fail with a permission error
        // For our normal shell test, it should succeed
        assert!(
            result.is_ok(),
            "Full path commands should work in normal shell"
        );
    }

    #[tokio::test]
    async fn test_format_output_ansi_handling() {
        let infra = Arc::new(MockInfrastructure::new());
        // Test with keep_ansi = true (should preserve ANSI codes)
        let ansi_output = CommandOutput {
            stdout: "\x1b[32mSuccess\x1b[0m".to_string(),
            stderr: "\x1b[31mWarning\x1b[0m".to_string(),
            command: "ls -la".into(),
            exit_code: Some(0),
        };
        let preserved = format_output(&infra, ansi_output, true, PREFIX_CHARS, SUFFIX_CHARS)
            .await
            .unwrap();
        insta::assert_snapshot!("format_output_ansi_preserved", preserved);

        // Test with keep_ansi = false (should strip ANSI codes)
        let ansi_output = CommandOutput {
            stdout: "\x1b[32mSuccess\x1b[0m".to_string(),
            stderr: "\x1b[31mWarning\x1b[0m".to_string(),
            command: "ls -la".into(),
            exit_code: Some(0),
        };
        let stripped = format_output(&infra, ansi_output, false, PREFIX_CHARS, SUFFIX_CHARS)
            .await
            .unwrap();
        insta::assert_snapshot!("format_output_ansi_stripped", stripped);
    }

    #[tokio::test]
    async fn test_format_output_with_large_command_output() {
        let infra = Arc::new(MockInfrastructure::new());
        // Using tiny PREFIX_CHARS and SUFFIX_CHARS values (30) to test truncation with
        // minimal content This creates very small snapshots while still testing
        // the truncation logic
        const TINY_PREFIX: usize = 30;
        const TINY_SUFFIX: usize = 30;

        // Create a test string just long enough to trigger truncation with our small
        // prefix/suffix values
        let test_string = "ABCDEFGHIJKLMNOPQRSTUVWXYZ".repeat(4); // 104 characters

        let ansi_output = CommandOutput {
            stdout: test_string.clone(),
            stderr: test_string,
            command: "ls -la".into(),
            exit_code: Some(0),
        };

        let preserved = format_output(&infra, ansi_output, false, TINY_PREFIX, TINY_SUFFIX)
            .await
            .unwrap();
        // Use a specific name for the snapshot instead of auto-generated name
        insta::assert_snapshot!(
            "format_output_large_command",
            TempDir::normalize(&preserved)
        );
    }
}
