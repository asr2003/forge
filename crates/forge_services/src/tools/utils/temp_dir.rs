use std::path::PathBuf;

use anyhow::Context;

pub struct TempDir {
    temp_dir: tempfile::TempDir,
}

impl TempDir {
    const START_MARKER: &'static str = "___START___";
    const END_MARKER: &'static str = "___END___";
    pub fn new() -> anyhow::Result<Self> {
        let temp_dir = Self::temp_dir()?;
        Ok(Self {
            temp_dir: tempfile::Builder::new()
                .prefix(Self::START_MARKER)
                .suffix(Self::END_MARKER)
                .tempdir_in(temp_dir.clone())
                .with_context(|| {
                    format!("failed to create temp directory in: {}", temp_dir.display())
                })?,
        })
    }

    pub fn path(&self) -> std::path::PathBuf {
        self.temp_dir.path().to_path_buf()
    }

    fn temp_dir() -> anyhow::Result<PathBuf> {
        Ok(std::env::temp_dir().canonicalize()?)
    }

    /// Use this function to remove the temporary directory path from the
    /// content
    pub fn normalize(input: &str) -> String {
        let pattern = format!(
            "{}/{}.*{}",
            Self::temp_dir().unwrap().display(),
            Self::START_MARKER,
            Self::END_MARKER
        );
        let pattern = regex::Regex::new(&pattern).unwrap();
        let content = pattern.replace_all(input, "[TEMP_DIR]");

        content.to_string()
    }
}
