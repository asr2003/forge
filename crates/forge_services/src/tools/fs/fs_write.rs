use std::fmt::Write;
use std::path::Path;
use std::sync::Arc;

use anyhow::Context;
use bytes::Bytes;
use console::strip_ansi_codes;
use forge_display::{DiffFormat, TitleFormat};
use forge_domain::{
    EnvironmentService, ExecutableTool, FSWriteInput, NamedTool, ToolCallContext, ToolDescription,
    ToolName,
};
use forge_tool_macros::ToolDescription;

use crate::tools::syn;
use crate::tools::utils::{assert_absolute_path, format_display_path};
use crate::{FsMetaService, FsReadService, FsWriteService, Infrastructure};

// Using FSWriteInput from forge_domain

/// Use it to create a new file at a specified path with the provided content.
/// Always provide absolute paths for file locations. The tool
/// automatically handles the creation of any missing intermediary directories
/// in the specified path.
/// IMPORTANT: DO NOT attempt to use this tool to move or rename files, use the
/// shell tool instead.
#[derive(ToolDescription)]
pub struct FSWrite<F>(Arc<F>);

impl<F: Infrastructure> FSWrite<F> {
    pub fn new(f: Arc<F>) -> Self {
        Self(f)
    }

    /// Formats a path for display, converting absolute paths to relative when
    /// possible
    ///
    /// If the path starts with the current working directory, returns a
    /// relative path. Otherwise, returns the original absolute path.
    fn format_display_path(&self, path: &Path) -> anyhow::Result<String> {
        // Get the current working directory
        let env = self.0.environment_service().get_environment();
        let cwd = env.cwd.as_path();

        // Use the shared utility function
        format_display_path(path, cwd)
    }
}

impl<F> NamedTool for FSWrite<F> {
    fn tool_name() -> ToolName {
        ToolName::new("forge_tool_fs_create")
    }
}

#[async_trait::async_trait]
impl<F: Infrastructure> ExecutableTool for FSWrite<F> {
    type Input = FSWriteInput;

    async fn call(&self, context: ToolCallContext, input: Self::Input) -> anyhow::Result<String> {
        // Validate absolute path requirement
        let path = Path::new(&input.path);
        assert_absolute_path(path)?;

        // Validate file content if it's a supported language file
        let syntax_warning = syn::validate(&input.path, &input.content);

        // Create parent directories if they don't exist
        if let Some(parent) = Path::new(&input.path).parent() {
            tokio::fs::create_dir_all(parent)
                .await
                .with_context(|| format!("Failed to create directories: {}", input.path))?;
        }

        // Check if the file exists
        let file_exists = self.0.file_meta_service().is_file(path).await?;

        // If file exists and overwrite flag is not set, return an error with the
        // existing content
        if file_exists && !input.overwrite {
            let existing_content = self.0.file_read_service().read_utf8(path).await?;
            return Err(anyhow::anyhow!(
                "File already exists at {}. If you need to overwrite it, set overwrite to true.\n\nExisting content:\n{}",
                input.path,
                existing_content
            ));
        }

        // record the file content before they're modified
        let old_content = if file_exists {
            // if file already exists, we should be able to read it.
            self.0.file_read_service().read_utf8(path).await?
        } else {
            // if file doesn't exist, we should record it as an empty string.
            "".to_string()
        };

        // Write file only after validation passes and directories are created
        self.0
            .file_write_service()
            .write(Path::new(&input.path), Bytes::from(input.content.clone()))
            .await?;

        let mut result = String::new();

        writeln!(result, "---")?;
        writeln!(result, "path: {}", &input.path)?;
        if file_exists {
            writeln!(result, "operation: OVERWRITE")?;
        } else {
            writeln!(result, "operation: CREATE")?;
        }
        writeln!(result, "total_chars: {}", input.content.len())?;
        if let Some(warning) = syntax_warning {
            writeln!(result, "Warning: {}", &warning.to_string())?;
        }
        writeln!(result, "---")?;

        // record the file content after they're modified
        let new_content = self.0.file_read_service().read_utf8(path).await?;
        let diff = DiffFormat::format(&old_content, &new_content);
        let title = if file_exists {
            writeln!(result, "{}", strip_ansi_codes(&diff))?;
            "Overwrite"
        } else {
            "Create"
        };

        // Use the formatted path for display
        let formatted_path = self.format_display_path(path)?;

        context
            .send_text(format!(
                "{}",
                TitleFormat::debug(title).sub_title(formatted_path)
            ))
            .await?;

        context.send_text(diff).await?;

        Ok(result)
    }
}

#[cfg(test)]
mod test {
    use std::path::Path;
    use std::sync::Arc;

    use insta::assert_snapshot;
    use pretty_assertions::assert_eq;

    use super::*;
    use crate::attachment::tests::MockInfrastructure;
    use crate::tools::utils::TempDir;
    use crate::{FsMetaService, FsReadService};

    async fn assert_path_exists(path: impl AsRef<Path>, infra: &MockInfrastructure) {
        assert!(
            infra
                .file_meta_service()
                .is_file(path.as_ref())
                .await
                .is_ok()
                || path.as_ref().exists(),
            "Path should exist"
        );
    }

    #[tokio::test]
    async fn test_fs_write_success() {
        let temp_dir = TempDir::new().unwrap();
        let file_path = temp_dir.path().join("test.txt");
        let content = "Hello, World!";

        let infra = Arc::new(MockInfrastructure::new());
        let fs_write = FSWrite::new(infra.clone());
        let output = fs_write
            .call(
                ToolCallContext::default(),
                FSWriteInput {
                    path: file_path.to_string_lossy().to_string(),
                    content: content.to_string(),
                    overwrite: false,
                },
            )
            .await
            .unwrap();

        // Normalize the output to remove temp directory paths
        let normalized_output = TempDir::normalize(&output);
        assert_snapshot!(normalized_output);

        // Verify file was actually written
        let content = infra
            .file_read_service()
            .read_utf8(&file_path)
            .await
            .unwrap();
        assert_eq!(content, "Hello, World!")
    }

    #[tokio::test]
    async fn test_fs_write_invalid_rust() {
        let temp_dir = TempDir::new().unwrap();
        let file_path = temp_dir.path().join("test.rs");

        let infra = Arc::new(MockInfrastructure::new());
        let fs_write = FSWrite::new(infra.clone());
        let result = fs_write
            .call(
                ToolCallContext::default(),
                FSWriteInput {
                    path: file_path.to_string_lossy().to_string(),
                    content: "fn main() { let x = ".to_string(),
                    overwrite: false,
                },
            )
            .await;

        let output = result.unwrap();
        // Normalize the output to remove temp directory paths
        let normalized_output = TempDir::normalize(&output);
        assert_snapshot!(normalized_output);
    }

    #[tokio::test]
    async fn test_fs_write_valid_rust() {
        let temp_dir = TempDir::new().unwrap();
        let file_path = temp_dir.path().join("test.rs");

        let infra = Arc::new(MockInfrastructure::new());
        let fs_write = FSWrite::new(infra.clone());
        let content = "fn main() { let x = 42; }";
        let result = fs_write
            .call(
                ToolCallContext::default(),
                FSWriteInput {
                    path: file_path.to_string_lossy().to_string(),
                    content: content.to_string(),
                    overwrite: false,
                },
            )
            .await;

        let output = result.unwrap();
        // Normalize the output to remove temp directory paths
        let normalized_output = TempDir::normalize(&output);
        assert_snapshot!(normalized_output);
        // Still keep basic assertions for specific conditions
        assert!(!output.contains("Warning:"));

        // Verify file contains valid Rust code
        let content = infra
            .file_read_service()
            .read_utf8(&file_path)
            .await
            .unwrap();
        assert_eq!(content, "fn main() { let x = 42; }");
    }

    #[tokio::test]
    async fn test_fs_write_single_directory_creation() {
        let temp_dir = TempDir::new().unwrap();
        let nested_path = temp_dir.path().join("new_dir").join("test.txt");
        let content = "Hello from nested file!";

        let infra = Arc::new(MockInfrastructure::new());
        let fs_write = FSWrite::new(infra.clone());
        let result = fs_write
            .call(
                ToolCallContext::default(),
                FSWriteInput {
                    path: nested_path.to_string_lossy().to_string(),
                    content: content.to_string(),
                    overwrite: false,
                },
            )
            .await
            .unwrap();

        // Normalize the output to remove temp directory paths
        let normalized_result = TempDir::normalize(&result);
        assert_snapshot!(normalized_result);

        // Verify both directory and file were created
        assert_path_exists(&nested_path, &infra).await;
        assert_path_exists(nested_path.parent().unwrap(), &infra).await;

        // Verify content
        let written_content = infra
            .file_read_service()
            .read_utf8(&nested_path)
            .await
            .unwrap();
        assert_eq!(written_content, content);
    }

    #[tokio::test]
    async fn test_fs_write_deep_directory_creation() {
        let temp_dir = TempDir::new().unwrap();
        let deep_path = temp_dir
            .path()
            .join("level1")
            .join("level2")
            .join("level3")
            .join("deep.txt");
        let content = "Deep in the directory structure";

        let infra = Arc::new(MockInfrastructure::new());
        let fs_write = FSWrite::new(infra.clone());
        let result = fs_write
            .call(
                ToolCallContext::default(),
                FSWriteInput {
                    path: deep_path.to_string_lossy().to_string(),
                    content: content.to_string(),
                    overwrite: false,
                },
            )
            .await
            .unwrap();

        // Normalize the output to remove temp directory paths
        let normalized_result = TempDir::normalize(&result);
        assert_snapshot!(normalized_result);

        // Verify entire path was created
        assert_path_exists(&deep_path, &infra).await;
        let mut current = deep_path.parent().unwrap();
        while current != temp_dir.path() {
            assert_path_exists(current, &infra).await;
            current = current.parent().unwrap();
        }

        // Verify content
        let written_content = infra
            .file_read_service()
            .read_utf8(&deep_path)
            .await
            .unwrap();
        assert_eq!(written_content, content);
    }

    #[tokio::test]
    async fn test_fs_write_with_different_separators() {
        let temp_dir = TempDir::new().unwrap();

        // Use forward slashes regardless of platform
        let path_str = format!("{}/dir_a/dir_b/file.txt", temp_dir.path().to_string_lossy());
        let content = "Testing path separators";

        let infra = Arc::new(MockInfrastructure::new());
        let fs_write = FSWrite::new(infra.clone());
        let result = fs_write
            .call(
                ToolCallContext::default(),
                FSWriteInput {
                    path: path_str,
                    content: content.to_string(),
                    overwrite: false,
                },
            )
            .await
            .unwrap();

        // Normalize the output to remove temp directory paths
        let normalized_result = TempDir::normalize(&result);
        assert_snapshot!(normalized_result);

        // Convert to platform path and verify
        let platform_path = Path::new(&temp_dir.path())
            .join("dir_a")
            .join("dir_b")
            .join("file.txt");

        assert_path_exists(&platform_path, &infra).await;

        // Verify content
        let written_content = infra
            .file_read_service()
            .read_utf8(&platform_path)
            .await
            .unwrap();
        assert_eq!(written_content, content);
    }

    #[tokio::test]
    async fn test_fs_write_relative_path() {
        let infra = Arc::new(MockInfrastructure::new());
        let fs_write = FSWrite::new(infra.clone());
        let result = fs_write
            .call(
                ToolCallContext::default(),
                FSWriteInput {
                    path: "relative/path/file.txt".to_string(),
                    content: "test content".to_string(),
                    overwrite: false,
                },
            )
            .await;

        assert!(result.is_err());
        assert!(result
            .unwrap_err()
            .to_string()
            .contains("Path must be absolute"));
    }

    #[tokio::test]
    async fn test_fs_write_no_overwrite() {
        let temp_dir = TempDir::new().unwrap();
        let file_path = temp_dir.path().join("test_overwrite.txt");
        let original_content = "Original content";

        let infra = Arc::new(MockInfrastructure::new());
        // First, create the file
        infra
            .file_write_service()
            .write(&file_path, Bytes::from(original_content))
            .await
            .unwrap();

        // Now attempt to write without overwrite flag
        let fs_write = FSWrite::new(infra.clone());
        let result = fs_write
            .call(
                ToolCallContext::default(),
                FSWriteInput {
                    path: file_path.to_string_lossy().to_string(),
                    content: "New content".to_string(),
                    overwrite: false,
                },
            )
            .await;

        // Should result in an error
        assert!(result.is_err());
        let error_msg = result.unwrap_err().to_string();

        // Error should mention overwrite flag
        assert!(error_msg.contains("set overwrite to true"));

        // Error should contain the original file content
        assert!(error_msg.contains(original_content));

        // Make sure the file wasn't changed
        let content = infra
            .file_read_service()
            .read_utf8(&file_path)
            .await
            .unwrap();
        assert_eq!(content, original_content);
    }

    #[tokio::test]
    async fn test_format_display_path() {
        let temp_dir = TempDir::new().unwrap();
        let file_path = temp_dir.path().join("test.txt");

        // Create a mock infrastructure with controlled cwd
        let infra = Arc::new(MockInfrastructure::new());
        let fs_write = FSWrite::new(infra);

        // Test with a mock path
        let display_path = fs_write.format_display_path(Path::new(&file_path));

        // Since MockInfrastructure has a fixed cwd of "/test",
        // and our temp path won't start with that, we expect the full path
        assert!(display_path.is_ok());
        assert_eq!(display_path.unwrap(), file_path.display().to_string());
    }

    #[tokio::test]
    async fn test_fs_write_with_overwrite() {
        let temp_dir = TempDir::new().unwrap();
        let file_path = temp_dir.path().join("test_overwrite.txt");
        let original_content = "Original content";
        let new_content = "New content";

        let infra = Arc::new(MockInfrastructure::new());
        // First, create the file
        infra
            .file_write_service()
            .write(&file_path, Bytes::from(original_content))
            .await
            .unwrap();

        // Now attempt to write with overwrite flag
        let fs_write = FSWrite::new(infra.clone());
        let result = fs_write
            .call(
                ToolCallContext::default(),
                FSWriteInput {
                    path: file_path.to_string_lossy().to_string(),
                    content: new_content.to_string(),
                    overwrite: true,
                },
            )
            .await;

        // Should be successful
        assert!(result.is_ok());
        let success_msg = result.unwrap();

        // Normalize the output to remove temp directory paths
        let normalized_msg = TempDir::normalize(&success_msg);
        assert_snapshot!(normalized_msg);

        // Verify file was actually overwritten
        let content = infra
            .file_read_service()
            .read_utf8(&file_path)
            .await
            .unwrap();
        assert_eq!(content, new_content);
    }
}
