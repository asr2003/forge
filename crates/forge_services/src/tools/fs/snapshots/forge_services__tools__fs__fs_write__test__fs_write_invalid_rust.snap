---
source: crates/forge_services/src/tools/fs/fs_write.rs
expression: normalized_output
snapshot_kind: text
---
---
path: [TEMP_DIR]/test.rs
operation: CREATE
total_chars: 20
Warning: Syntax error found in file with extension rs. Hint: Please retry in raw mode without HTML-encoding angle brackets.
---
